Add enhancement for IPI on PowerVC (#1811)
openshift/enhancements: enhancements/installer/openstack-for-power-vc-ipi.md

# openstack-for-power-vc-ipi

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for tech preview and GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> This document describes how the [OpenStack][openstack-website] provider can be leveraged to provide infrastructure to deploy

Reviewer: ```suggestion
This document describes how the [OpenStack](https://github.com/kubernetes-sigs/cluster-api-provider-openstack) provider can be leveraged to provide infrastructure to deploy
```

OpenShift on the [Power Virtualization Center][powervc-website] offering. Power Virtualization Center, or PowerVC, is an IBM

> offering that aims to provide customers simplified virtualization management for their on-premise datacenters. Because PowerVC

Reviewer: ```suggestion
offering that provides virtualization management for their on-premise datacenters. Because PowerVC
```

is based on OpenStack, we intend to heavily use the existing OpenStack provider by using the same paths in the installer that

> would be reached when specifying the openstack platform. The goal of this is to modify the OpenShift installer to enable

Reviewer: ```suggestion
The goal of this is to extend the OpenShift installer to enable
```

deployment to a PowerVC environment.

## Motivation

- Current deployments of OpenShift on Power on premise use UPI or the Assisted Intaller.
-- While effective, they require a lot of user intervention to deploy.
-- Scaling is more difficult when deploying via these methods.
- Customers are searching for a simplified method of deploying OpenShift into their existing PowerVC environments.
- IPI provides an opinionated and simplified deployment method.

> -- Because it is built on top of OpenStack, we can use the provider nearly as-is.

Reviewer: define it in the 'because it is'

- IPI provides an easy way to scale after deployment.

### Goals

> - Provide a way to install OpenShift on PowerVC infrastructure using the OpenShift installer's OpenStack provider.

Reviewer: ```suggestion
- Enable cluster-admin's to install an OpenShift cluster on PowerVC infrastructure
- Use the OpenShift installer's openstack provider code base to define the opinionated infrastructure
- Modularize the infrastructure differences so they are isolated from the openstack provider
```

### Non-Goals

> - Public cloud IPI. This is already implemented by the PowerVS provider. 

Reviewer: ```suggestion
- Replace or use the IBM Cloud PowerVS provider
- Implement an entirely new provider
```

## Proposal

> To leverage the PowerVC platform, we will make use of the [OpenStack Cluster API Provider][capo-website], or CAPO.

Reviewer: ```suggestion
To leverage the PowerVC platform, we will make use of the [OpenStack Cluster API Provider](https://github.com/openshift/enhancements/pull/1811/files) or CAPO.
```

> The OpenShift Installer will provision VMs in a specified PowerVC environment. These VMs will serve as bootstrap,

Reviewer: ```suggestion
The OpenShift Installer provisions VMs in a specified PowerVC group (availability_zone). These VMs will serve as bootstrap,
```

control plane, and compute nodes.

> The expected workflow would be as follows:

Reviewer: ```suggestion
The workflow is:
```

> 1. The installer will authenticate to PowerVC in the same manner as the OpenStack provider does today (clouds.yaml)

Reviewer: ```suggestion
1. The installer authenticates to PowerVC using the OpenStack provider (clouds.yaml)
```

2. The installer will import a boot image to PowerVC from IBM Cloud Object Storage. This is the same image used by the
PowerVS provider.
3. The installer will follow the same flow as the OpenStack provider.

> 4. TODO(mjturek): Describe the DHCP service challenges.

Reviewer: move to risks and mitigations? just say acquires dhcp address?

> 5. ROUGH: The cluster will come up without load balancers, and users will add load balancers

Reviewer: put this under risks? and you can explain the lack of lbs

### User Stories

- A user of this installer would like to see a cluster running in their PowerVC environment quickly after kicking off
the installer for a default size (3 master and 2 worker) cluster.

> - A user would like a scaling solution for their OpenShift cluster.

Reviewer: ```suggestion
- A user scales up the workers in the OpenShift cluster, creating new VMs in PowerVC.
- A user scales down the workers in the OpenShift cluster, destroying the VMs in PowerVC.
```

Author: Done

### API Extensions
None

### Implementation Details/Notes/Constraints

#### Features used by the OpenStack Provider that are Not Supported in PowerVC 
1. Security Groups
2. Allowed Address Pairs

### Risks and Mitigations

TODO

## Design Details

Basic Walkthrough:
1. The user creates a DHCP network in PowerVC.
2. The user prepares the clouds.yaml with access information to their PowerVC environment. 
3. The installer uploads an OVA from IBM Cloud COS and makes it bootable.
4. The installer hands the install over to CAPO.

[Power VS Reference Doc][power-vs-reference-doc]

### Installation Configuration
Here is a sample of an **install-config.yaml** for a PowerVC installation.

> ```yaml

Reviewer: ```suggestion

```yaml
```

apiVersion: v1
baseDomain: example.com
compute:
- architecture: ppc64le
  hyperthreading: Enabled
  name: worker
  platform:
    powervc:
      zones:
        - e980
  replicas: 3
controlPlane:
  architecture: ppc64le
  hyperthreading: Enabled
  name: master
  platform:
    powervc:
      zones:

>         - e980

Reviewer: ```suggestion
        - e980
```
maybe use a non-machine type to indicate it could have a mix of system types?

  replicas: 3
metadata:
  creationTimestamp: null
  name: powervc-cluster 
networking:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  machineNetwork:
  - cidr: 10.20.176.0/20
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
platform:
  powervc:
    loadBalancer:
      type: UserManaged
    apiVIPs:
    - 10.20.184.56
    cloud: powervc
    defaultMachinePlatform:
      type: project-name
    ingressVIPs:
    - 10.20.184.56
    machinesSubnet: 8f895eaa-d54f-4c2e-8f00-e1c5eeb51aa1
publish: External
pullSecret: ''

> sshKey: '

Reviewer: ```suggestion
sshKey: ''
```

```

### Networking
TODO: Call out DHCP prerequisites

#### Load Balancing
TODO: Call out LB post deploy tasks
#### DNS
TODO: Investigate how OpenStack handles this

### Red Hat Enterprise Linux CoreOS (RHCOS) OVA

> The RHCOS build pipeline publishes an OVA disk image to the IBM Cloud Object Storage (COS) buckets in each

Reviewer: ```suggestion

The RHCOS build pipeline publishes an OVA disk image to the IBM Cloud Object Storage (COS) buckets in each
```

preview is dicey when there are no returns

supported region. We will import the image to the desired Power VS instance, transforming it into an installable
boot image when it does so. The cluster-api-provider will later reference this same boot image to add workers.

### Open Questions

#### CI Plan
Currently, Multi-Arch CI for Power connects to dedicated hardware provided by IBM and provisions clusters using IPI for
libvirt, which isn't a supported customer configuration.

We will create prow jobs that uses dedicated hardware that hosts a PowerVC cluster.

### Test Plan

Multi-Arch currently tests results by running upgrade and e2e tests with the OpenShift Prow CI environment. Our test
plan would extend this to include these IPI scenarios. Additionally, new documentation will be needed for PowerVC,
and the test plan will incorporate review and validation of this.

### Graduation Criteria

> - Our initial target is to work towards a `Tech Preview`, then graduate towards future milestones. This is expected to

Reviewer: mention featuregate then tech preview then GA?

- [ ] End-user documentation
- [ ] CI e2e jobs created to cover installation against this platform
- [ ] Plan defined & support engaged to gather feedback from users

#### Tech Preview -> GA
- [ ] CI has been upgraded to cover testing of modifications to the installer (PR validation)
- [ ] CI "soak" for to ensure a stable baseline
- [ ] Final optimizations for install time
- [ ] Expanded testing (upgrade, downgrade, scale, load/performance)
- [ ] Documentation is finalized

#### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

### Upgrade / Downgrade Strategy

All upgrade and downgrade expectations are the same for existing IPI deployments.

### Version Skew Strategy

This should be the same as other IPI deployments.

### Operational Aspects of API Extensions
N/a

#### Failure Modes
N/a

#### Support Procedures
N/a

## Implementation History

Major milestones in the life cycle of a proposal should be tracked in `Implementation History`.

## Drawbacks

## Alternatives (Not Implemented)

We can continue to rely on our UPI playbooks but this is not ideal as they are not supported. We can also point users to

> our assisted installer support for PowerVC, but we believe users desire a more automated solution in this environment.

Reviewer 2: We need to explore `platform: external`: 
Is it feasible to use `platform: external` and rely back on OpenStack as a provider? 
Comparing this alternative to the proposal (overhead, maintenance, tests and validation, future features) can also help.

Author: addressed with a few sentences in the next commit

None

### Implementation Details/Notes/Constraints

While we can reuse most of what the openstack platform does as-is,
there are core differences that we'll need to address in the installer.
PowerVC handles images in a very specific way, so we'll need to perform
some extra steps to make the images bootable. Additionally, we'll need
to use swift to host ignition as glance is configured in a way that
we cannot guarantee the file can be downloaded.

### Risks and Mitigations

While the additional functionality will not pose a risk to other
teams, it will expand what is tested and supported by the IBM team. We are
confident that it will fit in our existing team so the risk is minimal.

### Drawbacks

There will be some prerequisites expected from the user, however this
is not unheard of in the installer.
 
### Implementation Details/Notes/Constraints
We will need to work around some features that the existing OpenStack
platform uses that we do not support. For example, we'll need to avoid
using security groups. We've proven this can be done.

#### Features used by the OpenStack Provider that are Not Supported in PowerVC 
1. Security Groups
2. Allowed Address Pairs

> 3. Native DHCP networks

Reviewer: Load Balancer?

## Design Details

Basic Walkthrough:
1. The user creates a DHCP network in PowerVC.
2. The user prepares the clouds.yaml with access information to their PowerVC environment. 

> 3. The installer uploads an OVA from IBM Cloud COS and makes it bootable.

Reviewer: Disconnected?

offering that provides virtualization management for their on-premise datacenters. Because PowerVC
is based on OpenStack, the new PowerVC platform heavily uses the same paths in the installer that
the OpenStack platform uses. The goal of this enhancement is to add a new powervc platform to the OpenShift
installer, that essentially serves as a shim provider of the existing openstack platform. This new
platform will acount for the differences between powervc and openstack to enable deployment of OpenShift to a PowerVC environment.

## Motivation

- Current deployments of OpenShift on Power on premise are user-provisioned, agent based, or use the assisted installer.
-- While effective, they require a lot of user intervention to deploy.
-- Automation around these features is not supported.
-- Scaling is more difficult when deploying via these methods.
- Customers are searching for a simplified method of deploying OpenShift into their existing PowerVC environments.
- IPI provides an opinionated and simplified deployment method.

> -- Because PowerVC is built on top of OpenStack, we can use the OpenStack IPI platform nearly as-is.

Reviewer 3: > Adding code paths specific to PowerVC to the OpenStack provider was explored

Also to add to this - even if this was the case, there would need to be some way in the install config to expose that it is PowerVC and not Openstack.

> -- Because PowerVC is built on top of OpenStack, we can use the OpenStack IPI platform nearly as-is.

Reviewer 3: Got it -so the main argument against using the existing Openstack platform and having additional options like disabling security group creation was not favored because we do not want to expose this as an option that is available for users outside of PowerVC?

> - Clusters deployed with IPI are equipped with an easy way to scale after deployment.

Reviewer 4: Note that it's possible to deploy the control plane with UPI but still use platform:openstack so that the worker nodes are created and scaled by the Machine API. Having an IPI installer is not a prerequisite for that.

Author: Good point... I'll add some clarity

- Create a "shim" platform around the existing OpenStack platform.
-- Modularize the infrastructure differences so they are isolated from the openstack provider

### Non-Goals

- Replace or use the IBM Cloud PowerVS provider. 
- Implement an entirely new provider.
-- Changes will be made to the installer, but we will call out to the same components as the OpenStack platform does.

## Proposal

To leverage the PowerVC platform, we will make use of the [OpenStack Cluster API Provider](https://github.com/openshift/cluster-api-provider-openstack), or CAPO.
The OpenShift Installer provisions VMs in a specified PowerVC environment. These VMs will serve as bootstrap,
control plane, and compute nodes.

### Workflow Description
The workflow is:
1. The installer authenticates to PowerVC in the same manner as the OpenStack provider does today (clouds.yaml)
2. The installer imports a boot image to PowerVC from IBM Cloud Object Storage. This is the same image used by the
PowerVS provider.
3. The installer follows the same flow as the OpenStack provider.
4. Because PowerVC does not have native DHCP network support, we are requiring users set up their own DHCP network service before installing.

> 5. The cluster will come up without load balancers. Cluster admins will add load balancers as a day-2 operation.

Reviewer 4: So it will or will not be using the on-prem networking?

### API Extensions
None

### User Stories
- A user of this installer would like to see a cluster running in their PowerVC environment quickly after kicking off

> the installer for a default size (3 control plane and 2 compute) cluster.

Author: @mandre you're right, we will be using external loadbalancers and the platform will most likely be hard coded to use them, unless PowerVC introduces a loadbalancing solution in the future.

Fixed the compute count as well

to use swift to host ignition as glance is configured in a way that
we cannot guarantee the file can be downloaded.

### Risks and Mitigations

While the additional functionality will not pose a risk to other
teams, it will expand what is tested and supported by the IBM team. We are
confident that it will fit in our existing team so the risk is minimal.

### Drawbacks

There will be some prerequisites expected from the user, however this
is not unheard of in the installer.
 
### Implementation Details/Notes/Constraints
We will need to work around some features that the existing OpenStack
platform uses that we do not support. For example, we'll need to avoid
using security groups. We've proven this can be done.

#### Features used by the OpenStack Provider that are Not Supported in PowerVC 
1. Security Groups
2. Allowed Address Pairs (Initial load balancing)
3. Native DHCP networks

## Design Details

Basic Walkthrough:
1. The user creates a DHCP network in PowerVC.
2. The user prepares the clouds.yaml with access information to their PowerVC environment. 
3. The installer uploads an OVA from IBM Cloud COS and makes it bootable.

> 4. The installer hands the install over to CAPO.

Author: we do not anticipate modifications to CAPO. I believe we can only support cinder.

### Installation Configuration
Here is a sample of an **install-config.yaml** for a PowerVC installation.

```yaml
apiVersion: v1
baseDomain: example.com
compute:
- architecture: ppc64le
  hyperthreading: Enabled
  name: worker
  platform:
    powervc:
      zones:
        - e980
  replicas: 3
controlPlane:
  architecture: ppc64le
  hyperthreading: Enabled
  name: master
  platform:
    powervc:
      zones:
        - e980
  replicas: 3
metadata:
  creationTimestamp: null
  name: powervc-cluster 
networking:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  machineNetwork:
  - cidr: 10.20.176.0/20
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
platform:
  powervc:
    loadBalancer:
      type: UserManaged
    apiVIPs:
    - 10.20.184.56
    cloud: powervc
    defaultMachinePlatform:
      type: project-name
    ingressVIPs:
    - 10.20.184.56
    machinesSubnet: 8f895eaa-d54f-4c2e-8f00-e1c5eeb51aa1
publish: External
pullSecret: ''
sshKey: ''
```

### Networking
Because PowerVC does not have native DHCP network support, we are requiring admins set up their own DHCP network service before installing.

#### Load Balancing

> The OpenStack provider uses the "allowed-address-pairs" feature of Neutron for initial loadbalancing. Because PowerVC lacks this feature, we expect admins to onboard their own load balancers as a day-2 operation.

Reviewer 4: Many questions about this.

A load balancer for the API is required during installation, so... how is that going to work if users don't set up a load balancer until day 2?

Author: @zaneb I was wrong about this, I had thought @hamzy's prototype came up without an LB but I was wrong.

I will clarify in the next revision. Yes, we'll need something to loadbalance

there are core differences that we'll need to address in the installer.
PowerVC handles images in a very specific way, so we'll need to perform
some extra steps to make the images bootable. Additionally, we'll need
to use swift to host ignition as glance is configured in a way that
we cannot guarantee the file can be downloaded.

### Risks and Mitigations

While the additional functionality will not pose a risk to other
teams, it will expand what is tested and supported by the IBM team. We are
confident that it will fit in our existing team so the risk is minimal.

### Drawbacks

There will be some prerequisites expected from the user, however this
is not unheard of in the installer.
 
### Implementation Details/Notes/Constraints
We will need to work around some features that the existing OpenStack
platform uses that we do not support. For example, we'll need to avoid
using security groups. We've proven this can be done.

#### Features used by the OpenStack Provider that are Not Supported in PowerVC 
1. Security Groups
2. Allowed Address Pairs (Initial load balancing)
3. Native DHCP networks

## Design Details

Basic Walkthrough:

> 1. The user creates a DHCP network in PowerVC.

Author: yes it is a requirement. There have been proposals to add static network support in the openstack provider which we might pursue in the future, but only if the openstack team wants it

## Drawbacks

## Alternatives (Not Implemented)

We can continue to rely on our UPI playbooks but this is not ideal as they are not supported. We can also point users to
our assisted installer support for PowerVC, but we believe users desire a more automated solution in this environment.

Adding code paths specific to PowerVC to the OpenStack provider was explored. For example, we tried adding the ability
to skip security group creation when using powervc. The OpenStack team was not open to accepting this as they didn't want
to imply that all OpenStack clusters should consider not using security groups. The changes needed to address the
differences between OpenStack and PowerVC would not benefit many users of the OpenStack IPI plaform. Because of this, we
decided that the best way to isolate these changes would be to create a shim provider that handled the differences in
environment while still leveraging the existing OpenStack IPI platform.

Using the external provider was another option that we explored. PowerVC is unique in that it is a variant of
OpenStack and would not need as much customization as the external provider enables. Additionally, we felt that the

> user experience would be easier if powervc was its own platform.

Reviewer 3: Also, I am not sure if we can use an exising in-tree provider (Openstack) and call it external.
